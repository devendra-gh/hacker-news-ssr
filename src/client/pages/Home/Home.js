import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import Head from '../../components/Head';
import { constant } from '../../constant';
import { setUserData } from '../../utils';
import TableList from '../../components/TableList';
import LineChart from '../../components/LineChart';
import Pagination from '../../components/Pagination';

const Home = (props) => {
  const {
    fetchHackerNews: loadHackerNews,
    updateUserData,
    home: { newsList, lineChart, pageNumber, isLoading },
  } = props;

  useEffect(() => {
    window.scrollTo(0, 0);
    loadHackerNews();
  }, [loadHackerNews]);

  const onClickNewsList = ({ objectID, hide, points }) => {
    const newUserData = {
      [objectID]: {
        hide,
        points,
      },
    };

    setUserData(newUserData);
    updateUserData({ newsList, newUserData, hide, objectID, points });
  };

  const onClickPrevNext = (e) => {
    const { name } = e.target;
    let currentPageNumber = pageNumber;

    if (name === constant.NEXT) {
      currentPageNumber += 1;
    } else {
      currentPageNumber -= 1;
    }

    loadHackerNews(currentPageNumber);
  };

  return (
    <div role="main" aria-label="Home Page">
      <Head />
      <div className="row">
        <div className="section">
          <TableList isLoading={isLoading} newsList={newsList} onClickNewsList={onClickNewsList} />
          <Pagination
            pageNumber={pageNumber}
            isLoading={isLoading}
            onClickPrevNext={onClickPrevNext}
          />
          {lineChart && lineChart.length ? <LineChart data={lineChart} /> : ''}
        </div>
      </div>
    </div>
  );
};

Home.propTypes = {
  home: PropTypes.shape({
    newsList: PropTypes.arrayOf(
      PropTypes.shape({
        objectID: PropTypes.string,
        num_comments: PropTypes.number,
        points: PropTypes.number,
        title: PropTypes.string,
        url: PropTypes.string,
        author: PropTypes.string,
        created_at: PropTypes.date,
      })
    ),
    // eslint-disable-next-line react/forbid-prop-types
    lineChart: PropTypes.arrayOf(PropTypes.any),
    pageNumber: PropTypes.number,
    isLoading: PropTypes.bool,
  }),
  fetchHackerNews: PropTypes.func,
  updateUserData: PropTypes.func,
};

Home.defaultProps = {
  home: {
    newsList: [],
    lineChart: [],
    pageNumber: 0,
    isLoading: false,
  },
  fetchHackerNews: null,
  updateUserData: null,
};

export default Home;
