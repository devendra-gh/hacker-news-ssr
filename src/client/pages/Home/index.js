import { connect } from 'react-redux';
import { baseAction, fetchHackerNews, updateUserData } from '../../actions';
import Home from './Home';

const loadData = (store) => {
  return store.dispatch(fetchHackerNews());
};

const mapStateToProps = (state) => {
  return {
    home: state.home,
  };
};

const mapDispatchToProps = (dispatch) => ({
  baseAction: (action, payload) => dispatch(baseAction(action, payload)),
  fetchHackerNews: (payload) => dispatch(fetchHackerNews(payload)),
  updateUserData: (payload) => dispatch(updateUserData(payload)),
  dispatch,
});

export default {
  component: connect(mapStateToProps, mapDispatchToProps)(Home),
  loadData,
};
