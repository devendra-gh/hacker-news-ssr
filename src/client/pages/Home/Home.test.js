import React from 'react';
import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Home from './Home';

describe('<Home />', () => {
  let props;

  beforeEach(() => {
    window.scrollTo = jest.fn();
    props = {
      home: {
        newsList: [],
        lineChart: [],
        pageNumber: 0,
        isLoading: false,
      },
      fetchHackerNews: jest.fn(),
      updateUserData: jest.fn(),
    };
  });

  afterEach(cleanup);

  test('Renders <Home /> component correctly', async () => {
    // eslint-disable-next-line react/jsx-props-no-spreading
    const { container, getByText } = render(<Home {...props} />);
    expect(container).toMatchSnapshot();
    expect(getByText('No Data Found')).toBeTruthy();
  });

  test('Renders <Home /> component correctly', async () => {
    props = {
      ...props,
      home: {
        ...props.home,
        newsList: [
          {
            objectID: '12345',
            num_comments: 10,
            points: 100,
            title: 'Name',
            url: 'google.com',
            author: 'Author',
            created_at: new Date('Thu Jul 20 2020 23:41:03'),
          },
        ],
        lineChart: [
          ['x', 'Vote'],
          ['12345', 100],
        ],
      },
    };
    // eslint-disable-next-line react/jsx-props-no-spreading
    const { container, getByText } = render(<Home {...props} />);
    expect(container).toMatchSnapshot();
    expect(getByText('10')).toBeTruthy();
    expect(getByText('100')).toBeTruthy();
    expect(getByText('Name')).toBeTruthy();
    expect(getByText('google.com')).toBeTruthy();
    expect(getByText('Author')).toBeTruthy();
    expect(getByText('hide')).toBeTruthy();
  });
});
