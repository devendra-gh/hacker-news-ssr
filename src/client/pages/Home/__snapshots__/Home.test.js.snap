// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> Renders <Home /> component correctly 1`] = `
<div>
  <div
    aria-label="Home Page"
    role="main"
  >
    <div
      class="row"
    >
      <div
        class="section"
      >
        <table
          aria-label="Hacker News List"
          class="striped responsive-table"
          role="contentinfo"
        >
          <thead>
            <tr>
              <th>
                Comments
              </th>
              <th>
                Vote count
              </th>
              <th>
                UpVote
              </th>
              <th>
                New Details
              </th>
              <th>
                Â 
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td
                colspan="5"
              >
                <div>
                  No Data Found
                </div>
              </td>
            </tr>
          </tbody>
        </table>
        <div
          class="carousel-container"
          style="padding-top: 20px; text-align: right;"
        >
          <button
            aria-label="Previous Button"
            class="btn red darken-4"
            disabled=""
            name="prev"
            role="navigation"
            style="margin-right: 10px;"
            type="button"
          >
            Previous
          </button>
          <button
            aria-label="Next Button"
            class="btn red darken-4"
            name="next"
            role="navigation"
            type="button"
          >
            Next
          </button>
        </div>
        
      </div>
    </div>
  </div>
</div>
`;
